rascal-vscode-extension/src/dap/RascalDebugView.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2018-2023, NWO-I CWI and Swat.engineering
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+import * as vscode from 'vscode';
+import { RascalDebugClient } from './RascalDebugClient';
+
+export class RascalDebugViewProvider implements vscode.TreeDataProvider<RascalReplNode> {
+    private changeEmitter = new vscode.EventEmitter<RascalReplNode | undefined>();
+    readonly onDidChangeTreeData = this.changeEmitter.event;
+
+    constructor(private readonly rascalDebugClient: RascalDebugClient, readonly context: vscode.ExtensionContext) {
+        vscode.window.onDidOpenTerminal(_e => {
+            this.changeEmitter.fire(undefined);
+        });
+        vscode.window.onDidCloseTerminal(_e => {
+            this.changeEmitter.fire(undefined);
+        });
+        vscode.window.onDidChangeActiveTerminal(_e => {
+            this.changeEmitter.fire(undefined);
+        });
+        vscode.debug.onDidStartDebugSession(_e => {
+            this.changeEmitter.fire(undefined);
+        });
+        vscode.debug.onDidTerminateDebugSession(_e => {
+            this.changeEmitter.fire(undefined);
+        });
+        vscode.debug.onDidReceiveDebugSessionCustomEvent(_e => {
+            this.changeEmitter.fire(undefined);
+        });
+        context.subscriptions.push(vscode.commands.registerCommand('rascalmpl.updateDebugView', () => {
+            this.changeEmitter.fire(undefined);
+        }));
```

Comment:
Can we make a small function that takes a any generic parameter, and just triggers an `changeEmitter.fire` event, and provide that as event handler for all these functions?

Also, don't all of these `onDid`... handlers need to be pushed to the subscribitions? to make sure we properly release resources on restart/update?

Reply:
(also this `rascalmpl.updateDebugView` is an example of a command not registered in the `package.json` ;)